src/container/linear.rs

```diff
@@ -0,0 +1,505 @@
+    Tiles, Tree,
+};
+
+// ----------------------------------------------------------------------------
+
+/// How large of a share of space each child has, on a 1D axis.
+///
+/// Used for [`Linear`] layouts (horizontal and vertical).
+#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+pub struct Shares {
+    /// How large of a share each child has.
+    ///
+    /// For instance, the shares `[1, 2, 3]` means that the first child gets 1/6 of the space,
+    /// the second gets 2/6 and the third gets 3/6.
+    shares: nohash_hasher::IntMap<TileId, f32>,
+}
+
+impl Shares {
+    pub fn replace_with(&mut self, remove: TileId, new: TileId) {
+        if let Some(share) = self.shares.remove(&remove) {
+            self.shares.insert(new, share);
+        }
+    }
+
+    /// Split the given width based on the share of the children.
+    pub fn split(&self, children: &[TileId], available_width: f32) -> Vec<f32> {
+        let mut num_shares = 0.0;
+        for &child in children {
+            num_shares += self[child];
+        }
+        if num_shares == 0.0 {
+            num_shares = 1.0;
+        }
+        children
+            .iter()
+            .map(|&child| available_width * self[child] / num_shares)
+            .collect()
+    }
+
+    pub fn retain(&mut self, keep: impl Fn(TileId) -> bool) {
+        self.shares.retain(|&child, _| keep(child));
+    }
+
+    pub fn sum(&self) -> f32 {
+        self.shares.values().sum()
+    }
+}
+
+impl<'a> IntoIterator for &'a Shares {
+    type Item = (&'a TileId, &'a f32);
+    type IntoIter = std::collections::hash_map::Iter<'a, TileId, f32>;
+
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        self.shares.iter()
+    }
+}
+
+impl std::ops::Index<TileId> for Shares {
+    type Output = f32;
+
+    #[inline]
+    fn index(&self, id: TileId) -> &Self::Output {
+        self.shares.get(&id).unwrap_or(&1.0)
+    }
+}
+
+impl std::ops::IndexMut<TileId> for Shares {
+    #[inline]
+    fn index_mut(&mut self, id: TileId) -> &mut Self::Output {
+        self.shares.entry(id).or_insert(1.0)
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+/// The direction of a [`Linear`] layout. Either horizontal or vertical.
+#[derive(
+    Clone, Copy, Debug, Default, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize,
+)]
+pub enum LinearDir {
+    #[default]
+    Horizontal,
+    Vertical,
+}
+
+/// Horizontal or vertical layout.
+#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+pub struct Linear {
+    pub children: Vec<TileId>,
+    pub dir: LinearDir,
+    pub shares: Shares,
+}
+
+impl Linear {
+    pub fn new(dir: LinearDir, children: Vec<TileId>) -> Self {
+        Self {
+            children,
+            dir,
+            ..Default::default()
+        }
+    }
+
+    pub fn new_binary(dir: LinearDir, children: [TileId; 2], fraction: f32) -> Self {
```

Comment:
should document that fraction is to be expected to be 0-1
could also call it `left_share_percentage`?